pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/BaseMultipleSegmentsConversionExecutor.java

```diff
@@ -352,6 +349,75 @@ public List<SegmentConversionResult> executeTask(PinotTaskConfig pinotTaskConfig
     }
   }
 
+  private int getThreadPoolSize(Map<String, String> taskConfigs) {
+    int nThreads = _minionConf.getProperty(Minion.SEGMENT_DOWNLOAD_THREAD_POOL_SIZE,
+        MinionConstants.DEFAULT_SEGMENT_DOWNLOAD_THREAD_POOL_SIZE);
+    nThreads = Integer.parseInt(taskConfigs.getOrDefault(MinionConstants.SEGMENT_DOWNLOAD_THREAD_POOL_SIZE,
+        String.valueOf(nThreads)));
+    return nThreads;
+  }
+
+  private void parallelDownloadAndUntarSegments(int nThreads, String tableNameWithType, String taskType,
+      String[] segmentNames, String[] downloadURLs, File tempDataDir, AtomicInteger recordCounter,
+      List<File> inputSegmentDirs)
+      throws Exception {
+
+    ExecutorService executorService = null;
+    int length = downloadURLs.length;
+    try {
+      executorService = Executors.newFixedThreadPool(nThreads);
+      List<Future<Void>> futures = new ArrayList<>(length);
+      for (int i = 0; i < length; i++) {
+        int index = i;
+        futures.add(executorService.submit(() -> {
+          downloadAndUntarSegment(tableNameWithType, taskType, segmentNames[index], downloadURLs[index],
+              tempDataDir, index, recordCounter, inputSegmentDirs);
+          return null;
+        }));
+
+        // Wait for all downloads to complete and cancel other tasks if any download fails
+        for (Future<Void> future : futures) {
```

Comment:
this should be outside the above for loop right ?

Reply:
+1, tests failures are related to this

Reply:
Thank you. This was a miss.